[Infrastructure] LocalNet on Kubernetes (#354)
pokt-network/pocket: docs/development/localnet-with-kubernetes.md

> # LocalNet on Kubernetes

Author: hackerman!

We are developing our own Kubernetes operator to manage v1 workloads both internally and for the community to make it easier to deploy pocket v1. While the operator is still in development, we can already utilize it to run our local networks locally on Kubernetes. This guide will show you how to do that.

- [LocalNet on Kubernetes](#localnet-on-kubernetes)
    - [Dependencies](#dependencies)
    - [Running the localnet](#running-the-localnet)
    - [Stopping / cleaning up the resources](#stopping--cleaning-up-the-resources)
    - [Interaction with the localnet](#interaction-with-the-localnet)
    - [How does it work?](#how-does-it-work)
    - [Troubleshooting](#troubleshooting)
    - [How to change configuration files](#how-to-change-configuration-files)

### Dependencies

* [tilt](https://docs.tilt.dev/install.html)

> * Kubernetes cluster ([different options available](https://docs.tilt.dev/choosing_clusters.html)), please make sure to run 1.23 version or older, until [this issue is resolved](https://github.com/zalando/postgres-operator/issues/2098) - otherwise Postgres DB is not going to be provisioned.

Author: Tilt now checks this

>   * `kubectl` is also required to be installed and configured to access the cluster, but that should happen automatically when you install kubernetes cluster locally.

Reviewer: Can we remove this line given the point above?

>   * `helm` is required to template the yaml manifests for the observability stack. Here are the instructions on how to install it: https://helm.sh/docs/intro/install/.

Reviewer: Let me know what you think of this style of writing for development docs. Personally, I don't think people like to read - just `copy-pasta` as fast as possible

```suggestion
  * `helm` - required to template the yaml manifests for the observability stack.
  	* Installation instructions: : https://helm.sh/docs/intro/install/.
```

> * pocket kubernetes operator codebase in `../pocket-operator` directory, relative to the pocket v1 codebase. If you don't have it, tilt will try to pull it from the `main` branch on first launch.

Reviewer: Why not make `pocket-operator a git submodule`?

Then we won't have a weird setup and it'll be easier to automate.

Author: From my experience, submodules are a huge PITA when used on repos where multiple people are collaborating. Though I agree, that would give us an ability to tie configuration files versions and logic from pocket-operator to the v1 repo. I just fear that managing a submodule might become a nightmare.

I mentioned that above - I won't be surprised if we remove pocket-operator from LocalNet and will use Kubernetes directly (with Tilt controlling the whole lifecycle). So I guess it doesn't matter if we use submodules or not - they are not going to be a deciding factor if operator sticks for LocalNet or not. :) 

Do you see any issues with using submodule? I can swap Tiltfile code with submodule.

  * Having this codebase available on your computer allows you to iterate/change the operator code while running the localnet.

### Running the localnet

> Start the LocalNet with `make localnet_up` command - this will start tilt. Tilt will prompt to press `space` to open a browser window with the UI where you can see the logs of all services running (including the operator, observability stack, etc.). You can also open the UI by going to `http://localhost:10350/` in your browser.

Reviewer: Alternative format for this section. Wdyt?
--- 

```
make localnet_up
```

The developer can then view the logs of all the services running (including the operator, observability stack, etc.) by either:
1. Pressing `space`
2. Going to [localhost:10350](http://localhost:10350/)

Author: Love it! ❤️

![tilt UI](tilt-ui.png)

### Stopping / cleaning up the resources

> `make localnet_down` will stop the localnet and clean up all the resources, except the postgres operator (in case you have other databases provisioned with it, the database itself will be destroyed).

Reviewer: If you do the following (for this command and others), the copy command will show up on github!

```
make localnet_down
```

### Interaction with the localnet

As the workloads run in Kubernetes, you can see and modify any resources on your local kubernetes by a tool of your choice (k9s, Lens, VSCode extension, etc.) - just be mindful that tilt and pocket-operator will change the resources back eventually, so you might want to disable tilt or turn off pocket-operator to make changes that you want to test.

We provide following make targets:

> `make localnet_shell` - opens a shell in the pod that has `client` cli available. It gets updated automatically whenever the code changes.

Reviewer: These are great comments and I think we should improve the Makefile to contain them. That's where developers will be looking!

`make localnet_client_debug` - opens a `client debug` cli. It allows to interact with blockchain, e.g. change pace maker mode, reset to genesis, etc. It gets updated automatically whenever the code changes.

> ### How does it work?

Reviewer: Wdyt of adding a mermaid diagram? https://mermaid-js.github.io/mermaid/#/

Stopping here for tonight. Will continue reviewing tomorrow

Author: That is a great idea, working on it.

Author: I initially built a larger diagram that was hard to make sense of, so I only kept the part with interaction between tilt, operator, and containers. Because that is the most tricky part.

### Troubleshooting

Sometimes you might experience issues running localnet on Kubernetes. They might be related to the fact different people run different clusters/OS/environments, and sometimes our laptops can go to sleep and that might not play well with virtual machines, postgres or pocket client.

- Check tilt web UI by pressing a space in the terminal where you started tilt or going to [this page](http://localhost:10350/) in your browser. If you see any errors, you can click "Trigger Update" on a resource that has issues to restart the service or retry a command.
- If triggering an update didn't help, please try to run `make localnet_down` and then `make localnet_up` again. This will clean up most of the resources and start the localnet from scratch.
- If `make localnet_down` didn't help, we suggest to rebuild local kubernetes cluster using the tool you're managing your cluster with - it could be Docker Desktop, Rancher Desktop, k3s, k3d, minikube, etc.
- Open an issue in this repo if you're still experiencing issues with running localnet using this guide.

### How to change configuration files

> LocalNet resources are managed by pocket-operator, so if you'd like to change a config file (e.g. genesis file), you can do it by modifying the corresponding CRD in `pocket-operator` repo.

Author: I think I should expand on that, but this should probably go into the pocket-operator documentation and linked here.
